Support Aurora point-in-time restore

Gufran opened the pull request:
Relates #5286

Changes proposed in this pull request:

* Added support for Aurora [point in time restore](https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_RestoreDBClusterToPointInTime.html) and clone.

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS="-run=TestAccAWSRDSCluster_PointInTimeRestore"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSRDSCluster_PointInTimeRestore -timeout 120m
=== RUN   TestAccAWSRDSCluster_PointInTimeRestore
=== PAUSE TestAccAWSRDSCluster_PointInTimeRestore
=== CONT  TestAccAWSRDSCluster_PointInTimeRestore
--- PASS: TestAccAWSRDSCluster_PointInTimeRestore (354.21s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       354.267s


```

Gufran commented:
@tombuildsstuff requesting review please.

papa99do commented:
Any update on this pull request? It would be great if it can be reviewed and included in the next release.

Gufran commented:
@papa99do I have no idea if this one is even scheduled for review yet. We are manually building this branch for internal usage until it is merged or there is some feedback.

ewbankkit commented:
[Cloning Aurora clusters cross-account](https://aws.amazon.com/about-aws/whats-new/2019/07/amazon_aurora_supportscloningacrossawsaccounts-/) uses this mechanism.
When this PR is merged we need to investigate the issue mentioned [here](https://github.com/terraform-providers/terraform-provider-aws/issues/10211#issuecomment-538045189).

chadmyers commented:
Is there any way to bump this up in priority to get it reviewed and merged?  Is there anything I can do to help. Thanks!

Gufran commented:
I'm sorry, I'm not sure if there is any way to bump it up or ask the maintainers to pick it up. It also looks like the changeset is no longer valid and I'll probably have to rethink the whole thing, that is if a maintainer is willing to review it.

chadmyers commented:
Dear Mr. Gufran, thank you for your contribution in this pull request. Above I was asking the maintainers if they could prioritize pulling in this pull request. I didn't mean to put more pressure on you to update the PR or anything like that. I was hoping the maintainers would look at this PR soon and I was volunteering to help if the PR needs updated since it has sat for so long.

Thank you, again, for this PR and also thank you to the maintainers for making these providers open source so we can collaborate on things like this!

anGie44 commented on aws/resource_aws_rds_cluster_test.go:
```diff
@@ -1651,6 +1677,54 @@ resource "aws_db_subnet_group" "test" {
+  cluster_identifier = "%s"
+  master_username = "root"
+  master_password = "password"
+  db_subnet_group_name = "${aws_db_subnet_group.test.name}"
```
I also found during testing that `db_subnet_group_name` isn't supported here (seeing: `Error: error creating RDS cluster: InvalidParameterValue: DB Subnet Group is not valid for DB Clusters`) so the vpc/subnet resources can be removed here to make the config a bit more concise

Gufran replied:
updated the config, also copying this from my other comment:

> db_subnet_group_name is a valid param according to the API docs, and also documented as existing functionality in terraform docs.
> I've also validated that it works fine in test. Do you think you might have hit another problem that caused terraform to show an error related to the subnet group param?

anGie44 commented on aws/resource_aws_rds_cluster_test.go:
```diff
@@ -1651,6 +1677,54 @@ resource "aws_db_subnet_group" "test" {
+  engine = "aurora-mysql"
+}
+
+resource "aws_vpc" "test" {
```
not necessary per comment above

Gufran replied:
`db_subnet_group_name` is a valid param according to the [API docs](https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_RestoreDBClusterToPointInTime.html), and also documented as existing functionality in [terraform docs](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/rds_cluster#db_subnet_group_name). 

I've also validated that it works fine in test. Do you think you might have hit another problem that caused terraform to show an error related to the subnet group param?

anGie44 replied:
hmm not sure what I was running into that first go around but test looks good atm 👍

anGie44 commented on aws/resource_aws_rds_cluster_test.go:
```diff
@@ -1651,6 +1677,54 @@ resource "aws_db_subnet_group" "test" {
+  }
+}
+
+resource "aws_subnet" "subnets" {
```
similar comment as above

anGie44 commented on aws/resource_aws_rds_cluster_test.go:
```diff
@@ -1651,6 +1677,54 @@ resource "aws_db_subnet_group" "test" {
+  }
+}
+
+resource "aws_db_subnet_group" "test" {
```
similar comment as above

anGie44 commented on aws/resource_aws_rds_cluster_test.go:
```diff
@@ -1651,6 +1677,54 @@ resource "aws_db_subnet_group" "test" {
+resource "aws_rds_cluster" "restored_pit" {
+	cluster_identifier = "%s"
+	skip_final_snapshot = true
+	engine = "aurora-mysql"
```
Nit: can reference the other resource here 👍 
```suggestion
	engine = aws_rds_cluster.test.engine
```

anGie44 reviewed (changes requested):
Hi @Gufran, thank you again for this PR and hope you can continue the work here to get it into a future release of the Provider soon! It's off to a great start, just a couple comments to help get the acceptance test in place and up-to-date with the current state of the provider that supports Terraform 0.12+. Let me know if you have any follow-up questions or if you are pressed for time, we are also happy to continue the work you have here ! (Please note, we generally provide contributors about a week's time to respond back with next steps)

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -203,12 +203,58 @@ func resourceAwsRDSCluster() *schema.Resource {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ForceNew:     true,
+							ValidateFunc: validateRdsRestorableTimeFormat,
```
In this PR, i've re-used  this method you have here but with a more generic name since it's applicable to other resources in the Provider: https://github.com/hashicorp/terraform-provider-aws/pull/15969/files#diff-78a0a97b5a4b0942c73e64eec6c3fbee6ce1e5e23744375b290543b2259ea320

Gufran replied:
Shall I rename the function here? I can't use the function from your PR because it is not merged yet.

And if you don't mind me asking, your PR is essentially the same as this. Why are you reviewing this one instead of continuing on that one? I am happy that you are reviewing this one and there is a chance it might go through, but I am also curious what's wrong with the other one?

anGie44 replied:
if you don't mind and agree with the renaming! once we get this PR merged, I can rebase. 

Nothing wrong with my PR (afik atm 😅 ), just needs a review from our internal team. I'm reviewing this one in the meantime so we can hopefully get the two enhancements in to close out #5286

Gufran replied:
@anGie44 renamed to `validateUTCTimestamp`.

Gufran commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -711,6 +757,126 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
 		}
 
+	} else if v, ok := d.GetOk("restore_in_time"); ok {
+		pointInTime := v.([]interface{})[0].(map[string]interface{})
+
+		restoreToTimeAttr := pointInTime["restore_to_time"].(string)
+		useLatestRestorableTimeAttr := pointInTime["use_latest_restorable_time"].(bool)
+
+		if useLatestRestorableTimeAttr && restoreToTimeAttr != "" {
+			return fmt.Errorf(`provider.aws: aws_rds_cluster: %s: Either "restore_to_time" or "use_latest_restorable_time" must be set`, d.Get("database_name").(string))
+		}
+
+		createOpts := &rds.RestoreDBClusterToPointInTimeInput{
+			DBClusterIdentifier:       aws.String(identifier),
+			DeletionProtection:        aws.Bool(d.Get("deletion_protection").(bool)),
+			SourceDBClusterIdentifier: aws.String(pointInTime["source_cluster_identifier"].(string)),
+			Tags:                      tags,
+		}
+
```
I think extracting it into a function only moves the code from this point to another without any additional benefit or reusability. The attribute `restore_to_point_in_time` get some safety guarantees from schema validations:

```
Type: schema.TypeList,
MaxItems: 1,
Elem: &schema.Resource{},
```

These three attributes together ensure that the attribute is decoded as a block and has at most 1 item. Further down the line `v, ok := d.GetOk("restore_to_point_in_time")` check ensures that the block is declared in code, and schema validation will prevent and empty value on it.

anGie44 replied:
fair point @Gufran !  instead of moving code that is only used in this part of the resource's create func, I would suggest then pulling the map values safely with the `v, ok := map[key].(string)` pattern (similar to what you have in L717 when extracting `restore_type`). this helps to ensure we don't hit unexpected panics and allows us to check if exactly 1 of the 2 possible attributes is set; we should be safe to assume that both `restore_to_time` and `use_latest_restorable_time` cannot be set at the same time as the `ConflictsWith` field is set on these attributes 👍 

e.g. 
```go
pointInTime := v.([]interface{})[0].(map[string]interface{})

if v, ok := pointInTime["restore_to_time"].(string); ok && v != "" {
  restoreToTime, err := time.Parse(time.RFC3339, restoreToTimeAttr)
  if err == nil {
    createOpts.RestoreToTime = aws.Time(restoreToTime)
  }
} 
if v, ok := pointInTime["use_latest_restorable_time"].(bool); ok && v {
  createOpts.UseLatestRestorableTime = aws.Bool(v)
}

if createOpts.RestoreToTime == nil && createOpts.UseLatestRestorableTime == nil {
  return fmt.Errorf(`provider.aws: aws_rds_cluster: %s: Either "restore_to_time" or "use_latest_restorable_time" must be set`, d.Get("database_name").(string))
}
```

anGie44 replied:
I should note though that the AWS RDS API is pretty good about notifying the user if one of these 2 fields is not set e.g. when neither param is sent 
```
Error: InvalidParameterValue: Please specify RestoreToTime or LatestRestorableTime
	status code: 400, 
```

Gufran commented:
Hi @anGie44, thanks for the review. I've incorporated latest changes into this branch, resolved the conflicts and also addressed the review comments. can you please take a look again? I'm working on the broken lint steps.

Gufran commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -389,7 +435,7 @@ func resourceAwsRDSCluster() *schema.Resource {
 			},
 
-			"tags": tagsSchema(),
+			"tags": tagsSchemaComputed(),
```
Its been so long I can't remember why I might have made that change. I've reverted it to use `tagsSchema` for now.

Gufran commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -711,6 +757,126 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+		err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+			resp, err := conn.RestoreDBClusterToPointInTime(createOpts)
+			if err != nil {
+				if isAWSErr(err, "DBClusterSnapshotNotFoundFault", "DBClusterSnapshotIdentifier doesn't refer to an existing DB cluster snapshot") {
```
Both of the errors are not retryable. I've removed the handler.

anGie44 commented:
sure thing @Gufran, thanks for your speedy reply! changes to address previous linting errors look good at the moment 👍

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,117 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+			log.Printf("[DEBUG]: RDS Cluster restore response: %s", resp)
+		}
+
+		if tfresource.TimedOut(err) {
+			_, err = conn.RestoreDBClusterToPointInTime(createOpts)
+		}
```
hi @Gufran -- with the `resource.Retry` now removed around the calls to `RestoreDBClusterToPointInTime` , this block is not needed; apologies for any confusion 😅 the `tfresource.TimedOut(err)` check is only applicable after the `retry` blocks

anGie44 commented on aws/resource_aws_rds_cluster_test.go:
```diff
@@ -364,6 +364,32 @@ func TestAccAWSRDSCluster_s3Restore(t *testing.T) {
+			{
+				Config: testAccAWSClusterConfig_pointInTimeRestoreSource(parentId, restoredId),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSClusterExists("aws_rds_cluster.test", &v),
+					resource.TestCheckResourceAttr("aws_rds_cluster.test", "engine", "aurora-mysql"),
+					testAccCheckAWSClusterExists("aws_rds_cluster.restored_pit", &c),
+					resource.TestCheckResourceAttr("aws_rds_cluster.restored_pit", "cluster_identifier", restoredId),
+					resource.TestCheckResourceAttr("aws_rds_cluster.test", "engine", "aurora-mysql"),
```
L386 duplicates L382..maybe "aws_rds_cluster.restored_pit" was meant in place? you could potentially change L386 or converge the 2 checks to one with `resource.TestCheckResourcePairAttr()`
```suggestion
					testAccCheckAWSClusterExists("aws_rds_cluster.test", &v),
					testAccCheckAWSClusterExists("aws_rds_cluster.restored_pit", &c),
					resource.TestCheckResourceAttr("aws_rds_cluster.restored_pit", "cluster_identifier", restoredId),
					resource.TestCheckResourceAttrPair("aws_rds_cluster.restored_pit", "engine", "aws_rds_cluster.test", "engine"),

```

anGie44 commented on aws/resource_aws_rds_cluster_test.go:
```diff
@@ -2634,6 +2660,54 @@ resource "aws_rds_cluster" "default" {
+  master_username      = "root"
+  master_password      = "password"
+  db_subnet_group_name = aws_db_subnet_group.test.name
+  skip_final_snapshot  = true
+  engine               = "aurora-mysql"
+}
+
+resource "aws_vpc" "test" {
+  cidr_block = "10.0.0.0/16"
+  tags = {
+    Name = "%s-vpc"
+  }
+}
+
+resource "aws_subnet" "subnets" {
+  count             = length(data.aws_availability_zones.available.names)
+  vpc_id            = aws_vpc.test.id
+  cidr_block        = "10.0.${count.index}.0/24"
+  availability_zone = data.aws_availability_zones.available.names[count.index]
+  tags = {
+    Name = "%s-subnet-${count.index}"
+  }
+}
+
+resource "aws_db_subnet_group" "test" {
+  name       = "%s-db-subnet-group"
+  subnet_ids = aws_subnet.subnets[*].id
+}
+
+resource "aws_rds_cluster" "restored_pit" {
+  cluster_identifier  = "%s"
+  skip_final_snapshot = true
+  engine              = "aurora-mysql"
+  restore_to_point_in_time {
+    source_cluster_identifier  = aws_rds_cluster.test.cluster_identifier
+    restore_type               = "full-copy"
+    use_latest_restorable_time = true
+  }
+}
+`, parentId, parentId, parentId, parentId, childId)
```
Small nitpick here to use formatter syntax to only require 1 reference of `parentId`
```suggestion
data "aws_availability_zones" "available" {}

resource "aws_rds_cluster" "test" {
  cluster_identifier   = "%[1]s"
  master_username      = "root"
  master_password      = "password"
  db_subnet_group_name = aws_db_subnet_group.test.name
  skip_final_snapshot  = true
  engine               = "aurora-mysql"
}

resource "aws_vpc" "test" {
  cidr_block = "10.0.0.0/16"
  tags = {
    Name = "%[1]s-vpc"
  }
}

resource "aws_subnet" "subnets" {
  count             = length(data.aws_availability_zones.available.names)
  vpc_id            = aws_vpc.test.id
  cidr_block        = "10.0.${count.index}.0/24"
  availability_zone = data.aws_availability_zones.available.names[count.index]
  tags = {
    Name = "%[1]s-subnet-${count.index}"
  }
}

resource "aws_db_subnet_group" "test" {
  name       = "%[1]s-db-subnet-group"
  subnet_ids = aws_subnet.subnets[*].id
}

resource "aws_rds_cluster" "restored_pit" {
  cluster_identifier  = "%s"
  skip_final_snapshot = true
  engine              = "aurora-mysql"
  restore_to_point_in_time {
    source_cluster_identifier  = aws_rds_cluster.test.cluster_identifier
    restore_type               = "full-copy"
    use_latest_restorable_time = true
  }
}
`, parentId, childId)
```

anGie44 commented on website/docs/r/rds_cluster.html.markdown:
```diff
@@ -190,6 +190,30 @@ resource "aws_rds_cluster" "example" {
+resource "aws_rds_cluster" "example-clone" {
+  # ... other configuration ...
+
+  restore_To_point_in_time {
```
```suggestion
  restore_to_point_in_time {
```

anGie44 replied:
I would also consider moving this entire block to below the `S3 Options` section to keep it close to that related section

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -217,12 +218,58 @@ func resourceAwsRDSCluster() *schema.Resource {
+							ConflictsWith: []string{"restore_to_point_in_time.0.restore_to_time"},
+						},
+
+						"restore_to_time": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ForceNew:     true,
+							ValidateFunc: validateUTCTimestamp,
+						},
```
to follow the current pattern in the provider, we should also add the field that conflicts w/this one:
```suggestion
						"restore_to_time": {
							Type:         schema.TypeString,
							Optional:     true,
							ForceNew:     true,
							ValidateFunc: validateUTCTimestamp,
							ConflictsWith: []string{"restore_to_point_in_time.0.use_latest_restorable_time"},
						},
```

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,117 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+		} else if restoreToTimeAttr != "" {
+			restoreToTime, _ := time.Parse(time.RFC3339, restoreToTimeAttr)
+			createOpts.RestoreToTime = aws.Time(restoreToTime)
+		} else {
```
In the event neither `restore_to_time` nor `use_latest_restorable_time` is set, we hit this `else` condition so the error message here is unfortunately not applicable .. see comment above for proposed refactor

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,117 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+			Tags:                      tags,
+		}
+
+		if attr, ok := pointInTime["restore_type"]; ok {
```
Nit: `Go` allows us cast the value here directly with `attr, ok := pointInTime["restore_type"].(string)`

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -217,12 +218,58 @@ func resourceAwsRDSCluster() *schema.Resource {
+						"use_latest_restorable_time": {
+							Type:          schema.TypeBool,
+							Optional:      true,
+							ForceNew:      true,
```
We can add the `Default:False` here as the API documents this value already defaults to `false`.

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,117 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+			createOpts.RestoreType = aws.String(attr.(string))
+		}
+
+		if useLatestRestorableTimeAttr {
+			createOpts.UseLatestRestorableTime = aws.Bool(useLatestRestorableTimeAttr)
+		} else if restoreToTimeAttr != "" {
+			restoreToTime, _ := time.Parse(time.RFC3339, restoreToTimeAttr)
+			createOpts.RestoreToTime = aws.Time(restoreToTime)
+		} else {
+			return fmt.Errorf(`provider.aws: aws_rds_cluster: %s: Both "restore_to_time" and "use_latest_restorable_time" cannot be set`, d.Get("database_name").(string))
+		}
```
In the event neither `restore_to_time` nor `use_latest_restorable_time` is set, we hit this else condition so the error message here is unfortunately no longer applicable .. see comment above for proposed refactor

anGie44 commented on website/docs/r/rds_cluster.html.markdown:
````diff
@@ -190,6 +190,30 @@ resource "aws_rds_cluster" "example" {
+}
+```
+
+* `source_cluster_identifier` - (Required) The identifier of the source database cluster from which to restore.
+* `restore_type` - (Optional) Type of restore to be performed.
+   Valid options are `full-copy` (default) and `copy-on-write`.
+* `use_latest_restorable_time` - (Optional) Set to true to restore the database cluster to the latest restorable backup time. Defaults to false.
+* `restore_to_time` - (Optional) Date and time in UTC format to restore the database cluster to.
+
+~> **NOTE:** `use_latest_restorable_time` and `restore_to_time` conflict with each other.
+
````
```suggestion
* `source_cluster_identifier` - (Required) The identifier of the source database cluster from which to restore.
* `restore_type` - (Optional) Type of restore to be performed.
   Valid options are `full-copy` (default) and `copy-on-write`.
* `use_latest_restorable_time` - (Optional) Set to true to restore the database cluster to the latest restorable backup time. Defaults to false. Conflicts with `restore_to_time`.
* `restore_to_time` - (Optional) Date and time in UTC format to restore the database cluster to. Conflicts with `use_latest_restorable_time`.
```

anGie44 commented:
Hi @Gufran -- just one additional follow-up regarding the acctest! I missed on first review that the `availability_zones` data source reference in the test needs to use this helper function `testAccAvailableAZsNoOptInConfig()` instead (per [docs](https://github.com/hashicorp/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md) section `Uses aws_availability_zones Data Source `). this will ensure the test passes in multiple regions. currently I'm seeing the test pass in `us-west-1` but erroring in `us-west-2`.

anGie44 commented on aws/resource_aws_rds_cluster_test.go:
```diff
@@ -2634,6 +2660,54 @@ resource "aws_rds_cluster" "default" {
 }
 
+func testAccAWSClusterConfig_pointInTimeRestoreSource(parentId, childId string) string {
+	return fmt.Sprintf(`
+data "aws_availability_zones" "available" {}
+
+resource "aws_rds_cluster" "test" {
+  cluster_identifier   = "%s"
+  master_username      = "root"
+  master_password      = "password"
+  db_subnet_group_name = aws_db_subnet_group.test.name
+  skip_final_snapshot  = true
+  engine               = "aurora-mysql"
+}
+
+resource "aws_vpc" "test" {
```
To address the region testing comment e.g. 
```suggestion
	return composeConfig(testAccAvailableAZsNoOptInConfig(), fmt.Sprintf(`
resource "aws_rds_cluster" "test" {
  cluster_identifier   = "%s"
  master_username      = "root"
  master_password      = "password"
  db_subnet_group_name = aws_db_subnet_group.test.name
  skip_final_snapshot  = true
  engine               = "aurora-mysql"
}

resource "aws_vpc" "test" {
```

Gufran commented:
@anGie44 everything done. can you take a look again please?

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,109 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+			Tags:                      tags,
+		}
+
+		if v, ok := pointInTime["restore_in_time"].(string); ok {
```
small typo: 
```suggestion
		if v, ok := pointInTime["restore_to_time"].(string); ok {
```

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,109 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+		log.Printf("[DEBUG] RDS Cluster restore options: %s", createOpts)
+
+		resp, err := conn.RestoreDBClusterToPointInTime(createOpts)
+		if err != nil {
+			log.Printf("[DEBUG]: RDS Cluster restore response: %s", resp)
+		}
```
should this be removed in favor of the `err` check below?

Gufran replied:
The debug log dumps the API response that isn't printed in the error log. I've merge the two conditions to only check for error once and kept both the log lines.

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,109 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+			Tags:                      tags,
+		}
+
+		if v, ok := pointInTime["restore_to_time"].(string); ok {
```
we'll still need to check that the string here is not empty e.g.
```
if v, ok := pointInTime["restore_to_time"].(string); ok && v != ""{
```
since by default this `pointInTime` map stores the value here as the empty string even when this value is not explicitly configured in Terraform.

This line change here should clear up the current test failure that's returning:
```
  TestAccAWSRDSCluster_PointInTimeRestore: resource_aws_rds_cluster_test.go:374: Step 1/1 error: Error running apply:
        Error: InvalidParameterCombination: You cannot specify both UseLatestRestorableTime and RestoreToTime at same time.
        	status code: 400,
```

anGie44 commented on website/docs/r/rds_cluster.html.markdown:
```diff
@@ -162,6 +162,28 @@ resource "aws_rds_cluster" "db" {
+* `source_cluster_identifier` - (Required) The identifier of the source database cluster from which to restore.
+* `restore_type` - (Optional) Type of restore to be performed.
+   Valid options are `full-copy` (default) and `copy-on-write`.
+* `use_latest_restorable_time` - (Optional) Set to true to restore the database cluster to the latest restorable backup time. Defaults to false.
+* `restore_to_time` - (Optional) Date and time in UTC format to restore the database cluster to.
```
Nit: to clarify the plan-time validation
```suggestion
* `use_latest_restorable_time` - (Optional) Set to true to restore the database cluster to the latest restorable backup time. Defaults to false. Conflicts with `restore_to_time`.
* `restore_to_time` - (Optional) Date and time in UTC format to restore the database cluster to. Conflicts with `use_latest_restorable_time`.
```

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,109 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+			createOpts.RestoreType = aws.String(attr)
+		}
+
+		if err := validateRestoreToPointInTimeInput(d); err != nil {
+			return err
+		}
```
looking this over in more detail, I think this validation might hinder operators' experiences as the values checked w/in this method can technically be added without any issue to the cluster creation, given they match the original cluster's values i.e.
```
resource "aws_rds_cluster" "restored_pit" {
  cluster_identifier  = "%s"
  skip_final_snapshot = true
  engine              = aws_rds_cluster.test.engine
  database_name       = aws_rds_cluster.test.database_name
  master_username     = aws_rds_cluster.test.master_username
  storage_encrypted   = aws_rds_cluster.test.storage_encrypted
  source_region       = aws_rds_cluster.test.source_region
  restore_to_point_in_time {
    source_cluster_identifier  = aws_rds_cluster.test.cluster_identifier
    restore_type               = "full-copy"
    use_latest_restorable_time = true
  }
}
```
To account for the possible case where values provided do not match the original cluster, resulting in a perpetual diff after `apply`, I would suggest documenting the expected behavior in the docs

anGie44 commented on website/docs/r/rds_cluster.html.markdown:
```diff
@@ -162,6 +162,28 @@ resource "aws_rds_cluster" "db" {
 This will not recreate the resource if the S3 object changes in some way. It's only used to initialize the database. This only works currently with the aurora engine. See AWS for currently supported engines and options. See [Aurora S3 Migration Docs](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/AuroraMySQL.Migrating.ExtMySQL.html#AuroraMySQL.Migrating.ExtMySQL.S3).
 
+### restore_to_point_in_time Argument Reference
+
```
```suggestion

~> **NOTE:**  The DB cluster is created from the source DB cluster with the same configuration as the original DB cluster, except that the new DB cluster is created with the default DB security group. Thus, the following arguments cannot be modified: `database_name`, `master_username`, `storage_encrypted`, `replication_source_identifier`, and `source_region`.
```

anGie44 commented on website/docs/r/rds_cluster.html.markdown:
```diff
@@ -125,6 +125,7 @@ The following arguments are supported:
 * `preferred_backup_window` - (Optional) The daily time range during which automated backups are created if automated backups are enabled using the BackupRetentionPeriod parameter.Time in UTC. Default: A 30-minute window selected at random from an 8-hour block of time per region. e.g. 04:00-09:00
 * `preferred_maintenance_window` - (Optional) The weekly time range during which system maintenance can occur, in (UTC) e.g. wed:04:00-wed:04:30
 * `replication_source_identifier` - (Optional) ARN of a source DB cluster or DB instance if this DB cluster is to be created as a Read Replica. If DB Cluster is part of a Global Cluster, use the [`lifecycle` configuration block `ignore_changes` argument](/docs/configuration/resources.html#ignore_changes) to prevent Terraform from showing differences for this argument instead of configuring this v ...
+* `restore_to_point_in_time` - (Optional) Nested attribute for [point in time restore](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_PIT.html). More details below.
```
let's reference the rds cluster dev guide page here to not confuse with the DB Instance restore
```suggestion
* `restore_to_point_in_time` - (Optional) Nested attribute for [point in time restore](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/USER_PIT.html). More details below.
```

Gufran commented:
@anGie44 Thank you for your patience and thorough review. I've made the changes, can you take a look once again?

anGie44 commented on aws/resource_aws_rds_cluster.go:
```diff
@@ -650,6 +697,102 @@ func resourceAwsRDSClusterCreate(d *schema.ResourceData, meta interface{}) error
+		log.Printf("[DEBUG] RDS Cluster restore options: %s", createOpts)
+
+		resp, err := conn.RestoreDBClusterToPointInTime(createOpts)
+		if err != nil {
+			log.Printf("[DEBUG]: RDS Cluster restore response: %s", resp)
+			log.Printf("[ERROR] Error restoring RDS Cluster: %s", err)
+			return err
+		}
```
one last change here and we'll be good to merge 👍  the `resp` obj here will most likely be `nil` on error so we should instead capture the `resp` on success like seen @ L916 below
```suggestion
		if err != nil {
			log.Printf("[ERROR] Error restoring RDS Cluster: %s", err)
			return err
		}
		
		log.Printf("[DEBUG]: RDS Cluster restore response: %s", resp)
```

Gufran commented:
@anGie44 all done.

anGie44 reviewed (approved):
Thanks for the fantastic work here @Gufran ! 🚀 

Output of acceptance tests:
```
--- PASS: TestAccAWSRDSClusterInstance_MonitoringRoleArn_RemovedToEnabled (579.64s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsEnabled_AuroraPostgresql (682.82s)
--- PASS: TestAccAWSRDSClusterInstance_generatedName (697.52s)
--- PASS: TestAccAWSRDSClusterInstance_PubliclyAccessible (712.88s)
--- PASS: TestAccAWSRDSClusterInstance_namePrefix (717.19s)
--- PASS: TestAccAWSRDSClusterInstance_CopyTagsToSnapshot (723.80s)
--- PASS: TestAccAWSRDSClusterInstance_disappears (746.74s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsEnabled_AuroraMysql1 (767.86s)
--- PASS: TestAccAWSRDSClusterInstance_isAlreadyBeingDeleted (783.13s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraMysql1 (788.66s)
--- PASS: TestAccAWSRDSClusterInstance_az (796.07s)
--- PASS: TestAccAWSRDSClusterInstance_kmsKey (798.91s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraMysql1_DefaultKeyToCustomKey (810.22s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsEnabled_AuroraMysql2 (878.27s)
--- PASS: TestAccAWSRDSCluster_AvailabilityZones (122.85s)
--- PASS: TestAccAWSRDSCluster_basic (172.80s)
--- PASS: TestAccAWSRDSCluster_missingUserNameCausesError (4.03s)
--- PASS: TestAccAWSRDSCluster_ClusterIdentifierPrefix (122.71s)
--- PASS: TestAccAWSRDSCluster_DbSubnetGroupName (136.82s)
--- PASS: TestAccAWSRDSCluster_BacktrackWindow (180.84s)
--- PASS: TestAccAWSRDSClusterInstance_MonitoringRoleArn_EnabledToRemoved (978.59s)
--- PASS: TestAccAWSRDSClusterInstance_MonitoringRoleArn_EnabledToDisabled (993.53s)
--- PASS: TestAccAWSRDSCluster_generatedName (122.15s)
--- PASS: TestAccAWSRDSClusterEndpoint_tags (1018.87s)
--- PASS: TestAccAWSRDSCluster_Tags (139.97s)
--- PASS: TestAccAWSRDSCluster_takeFinalSnapshot (153.69s)
--- PASS: TestAccAWSRDSCluster_EnabledCloudwatchLogsExports_Postgresql (132.65s)
--- PASS: TestAccAWSRDSCluster_updateIamRoles (135.87s)
--- PASS: TestAccAWSRDSCluster_EnabledCloudwatchLogsExports_MySQL (210.80s)
--- PASS: TestAccAWSRDSCluster_kmsKey (144.68s)
--- PASS: TestAccAWSRDSClusterEndpoint_basic (1125.17s)
--- PASS: TestAccAWSRDSCluster_encrypted (164.16s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraMysql2 (590.55s)
--- PASS: TestAccAWSRDSCluster_iamAuth (152.46s)
--- PASS: TestAccAWSRDSCluster_PointInTimeRestore (388.55s)
--- PASS: TestAccAWSRDSCluster_copyTagsToSnapshot (211.31s)
--- PASS: TestAccAWSRDSCluster_DeletionProtection (161.76s)
--- PASS: TestAccAWSRDSCluster_backupsUpdate (211.05s)
--- PASS: TestAccAWSRDSCluster_EngineMode_ParallelQuery (143.36s)
--- PASS: TestAccAWSRDSClusterInstance_MonitoringInterval (1270.13s)
--- PASS: TestAccAWSRDSCluster_EngineMode_Global (153.41s)
--- PASS: TestAccAWSRDSCluster_EngineMode_Multimaster (157.86s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_EngineMode_Global (137.26s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_EngineMode_Global_Add (139.26s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraPostgresql (660.89s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_EngineMode_Provisioned (125.73s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_EngineMode_Global_Update (150.72s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_EngineMode_Global_Remove (166.67s)
--- SKIP: TestAccAWSRDSCluster_SnapshotIdentifier_EngineMode_Serverless (0.00s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraPostgresql_DefaultKeyToCustomKey (676.95s)
--- PASS: TestAccAWSRDSClusterInstance_PerformanceInsightsKmsKeyId_AuroraMysql2_DefaultKeyToCustomKey (736.23s)
--- PASS: TestAccAWSRDSClusterInstance_CACertificateIdentifier (710.77s)
--- PASS: TestAccAWSRDSClusterInstance_basic (1512.95s)
--- PASS: TestAccAWSRDSCluster_Port (251.93s)
--- PASS: TestAccAWSRDSCluster_EngineVersion (436.53s)
--- PASS: TestAccAWSRDSCluster_EngineMode (517.84s)
--- PASS: TestAccAWSRDSCluster_ScalingConfiguration (373.76s)
--- PASS: TestAccAWSRDSCluster_ScalingConfiguration_DefaultMinCapacity (335.35s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EngineVersion_Different (340.80s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EngineVersion_Equal (350.62s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EngineMode_Provisioned (381.29s)
--- PASS: TestValidateTypeStringNullableBoolean (0.00s)
--- PASS: TestValidateTypeStringNullableFloat (0.00s)
--- PASS: TestValidateCloudWatchDashboardName (0.00s)
--- PASS: TestValidateCloudWatchEventRuleName (0.00s)
--- PASS: TestValidateLambdaFunctionName (0.00s)
--- PASS: TestValidateLambdaQualifier (0.00s)
--- PASS: TestValidateLambdaPermissionAction (0.00s)
--- PASS: TestValidateLambdaPermissionEventSourceToken (0.00s)
--- PASS: TestValidateAwsAccountId (0.00s)
--- PASS: TestValidateArn (0.00s)
--- PASS: TestValidateEC2AutomateARN (0.00s)
--- PASS: TestValidatePolicyStatementId (0.00s)
--- PASS: TestValidateCIDRNetworkAddress (0.00s)
--- PASS: TestValidateCIDRBlock (0.00s)
--- PASS: TestValidateIpv4CIDRBlock (0.00s)
--- PASS: TestValidateIpv6CIDRBlock (0.00s)
--- PASS: TestCidrBlocksEqual (0.00s)
--- PASS: TestCanonicalCidrBlock (0.00s)
--- PASS: TestValidateLogMetricFilterName (0.00s)
--- PASS: TestValidateLogMetricTransformationName (0.00s)
--- PASS: TestValidateLogGroupName (0.00s)
--- PASS: TestValidateLogGroupNamePrefix (0.00s)
--- PASS: TestValidateS3BucketLifecycleTimestamp (0.00s)
--- PASS: TestValidateSagemakerName (0.00s)
--- PASS: TestValidateDbEventSubscriptionName (0.00s)
--- PASS: TestValidateIAMPolicyJsonString (0.00s)
--- PASS: TestValidateStringIsJsonOrYaml (0.00s)
--- PASS: TestValidateSQSQueueName (0.00s)
--- PASS: TestValidateSQSFifoQueueName (0.00s)
--- PASS: TestValidateOnceAWeekWindowFormat (0.00s)
--- PASS: TestValidateOnceADayWindowFormat (0.00s)
--- PASS: TestValidateEcsPlacementConstraint (0.00s)
--- PASS: TestValidateEcsPlacementStrategy (0.00s)
--- PASS: TestValidateStepFunctionStateMachineName (0.00s)
--- PASS: TestValidateEmrCustomAmiId (0.00s)
--- PASS: TestValidateDmsEndpointId (0.00s)
--- PASS: TestValidateDmsCertificateId (0.00s)
--- PASS: TestValidateDmsReplicationInstanceId (0.00s)
--- PASS: TestValidateDmsReplicationSubnetGroupId (0.00s)
--- PASS: TestValidateDmsReplicationTaskId (0.00s)
--- PASS: TestValidateAccountAlias (0.00s)
--- PASS: TestValidateIamRoleProfileName (0.00s)
--- PASS: TestValidateIamRoleProfileNamePrefix (0.00s)
--- PASS: TestValidateApiGatewayUsagePlanQuotaSettings (0.00s)
--- PASS: TestValidateDocDBIdentifier (0.00s)
--- PASS: TestValidateElbName (0.00s)
--- PASS: TestValidateElbNamePrefix (0.00s)
--- PASS: TestValidateNeptuneEventSubscriptionName (0.00s)
--- PASS: TestValidateNeptuneEventSubscriptionNamePrefix (0.00s)
--- PASS: TestValidateDbSubnetGroupName (0.00s)
--- PASS: TestValidateNeptuneSubnetGroupName (0.00s)
--- PASS: TestValidateDbSubnetGroupNamePrefix (0.00s)
--- PASS: TestValidateNeptuneSubnetGroupNamePrefix (0.00s)
--- PASS: TestValidateDbOptionGroupName (0.00s)
--- PASS: TestValidateDbOptionGroupNamePrefix (0.00s)
--- PASS: TestValidateDbParamGroupName (0.00s)
--- PASS: TestValidateOpenIdURL (0.00s)
--- PASS: TestValidateAwsKmsName (0.00s)
--- PASS: TestValidateAwsKmsGrantName (0.00s)
--- PASS: TestValidateCognitoIdentityPoolName (0.00s)
--- PASS: TestValidateCognitoProviderDeveloperName (0.00s)
--- PASS: TestValidateCognitoSupportedLoginProviders (0.00s)
--- PASS: TestValidateCognitoIdentityProvidersClientId (0.00s)
--- PASS: TestValidateCognitoIdentityProvidersProviderName (0.00s)
--- PASS: TestValidateCognitoUserPoolEmailVerificationMessage (0.01s)
--- PASS: TestValidateCognitoUserPoolEmailVerificationSubject (0.00s)
--- PASS: TestValidateCognitoUserPoolSmsAuthenticationMessage (0.00s)
--- PASS: TestValidateCognitoUserPoolSmsVerificationMessage (0.00s)
--- PASS: TestValidateWafMetricName (0.00s)
--- PASS: TestValidateAwsSSMName (0.00s)
--- PASS: TestValidateBatchName (0.00s)
--- PASS: TestValidateBatchPrefix (0.00s)
--- PASS: TestValidateCognitoRoleMappingsAmbiguousRoleResolutionAgainstType (0.00s)
--- PASS: TestValidateCognitoRoleMappingsRulesConfiguration (0.00s)
--- PASS: TestValidateSecurityGroupRuleDescription (0.00s)
--- PASS: TestValidateCognitoRoles (0.00s)
--- PASS: TestValidateKmsKey (0.00s)
--- PASS: TestResourceAWSElastiCacheReplicationGroupAuthTokenValidation (0.00s)
--- PASS: TestValidateCognitoUserGroupName (0.00s)
--- PASS: TestValidateCognitoUserPoolId (0.00s)
--- PASS: TestValidateAmazonSideAsn (0.00s)
--- PASS: TestValidate4ByteAsn (0.00s)
--- PASS: TestValidateLaunchTemplateName (0.00s)
--- PASS: TestValidateLaunchTemplateId (0.00s)
--- PASS: TestValidateNeptuneParamGroupName (0.00s)
--- PASS: TestValidateNeptuneParamGroupNamePrefix (0.00s)
--- PASS: TestValidateCloudFrontPublicKeyName (0.00s)
--- PASS: TestValidateCloudFrontPublicKeyNamePrefix (0.00s)
--- PASS: TestValidateDxConnectionBandWidth (0.00s)
--- PASS: TestValidateLbTargetGroupName (0.00s)
--- PASS: TestValidateLbTargetGroupNamePrefix (0.00s)
--- PASS: TestValidateSecretManagerSecretName (0.00s)
--- PASS: TestValidateSecretManagerSecretNamePrefix (0.00s)
--- PASS: TestValidateRoute53ResolverName (0.00s)
--- PASS: TestCloudWatchEventCustomEventBusName (0.00s)
--- PASS: TestValidateServiceDiscoveryNamespaceName (0.00s)
--- PASS: TestValidateUTCTimestamp (0.00s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EngineMode_ParallelQuery (401.03s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_MasterPassword (370.98s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier (461.46s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_KmsKeyId (381.59s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_DeletionProtection (460.52s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_PreferredBackupWindow (361.30s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_PreferredMaintenanceWindow (350.69s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_MasterUsername (461.39s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_Tags (360.85s)
--- PASS: TestAccAWSRDSCluster_AllowMajorVersionUpgrade (1243.59s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_VpcSecurityGroupIds_Tags (364.50s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_VpcSecurityGroupIds (402.50s)
--- PASS: TestAccAWSRDSCluster_SnapshotIdentifier_EncryptedRestore (381.50s)
--- PASS: TestAccAWSRDSCluster_EnableHttpEndpoint (373.70s)
--- PASS: TestAccAWSRDSCluster_EngineVersionWithPrimaryInstance (1141.53s)
--- PASS: TestAccAWSRDSCluster_ReplicationSourceIdentifier_KmsKeyId (1512.03s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_ReplicationSourceIdentifier (1901.34s)
--- PASS: TestAccAWSRDSCluster_GlobalClusterIdentifier_PrimarySecondaryClusters (1980.75s)
```
```
--- FAIL: TestAccAWSRDSCluster_s3Restore (1589.99s) -- known failure reported in https://github.com/hashicorp/terraform-provider-aws/issues/13391
```
